internal/timecmp/cmp.go

```diff
@@ -0,0 +1,98 @@
+// and then stored on an entity as a metav1.Time object, future comparisons might not behave as
+// expected since the latter value will be truncated.
+//
+// The comparison functions provided by this package normalize values to the lowest granularity
+// of the values being compared before performing the actual comparison.
+package timecmp
+
+import (
+	"fmt"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// commonTime is the common interface between apiserver + Go stdlib time types necessary for
+// normalization.
+type commonTime interface {
+	Truncate(duration time.Duration) time.Time
+}
+
+// Equal returns true of the normalized versions of a and b are equal.
+//
+// Values are normalized to the lowest granularity between the two values: seconds if either
+// is metav1.Time, microseconds if either is metav1.MicroTime, or monotonically-stripped if
+// both are time.Time.
+func Equal(a, b commonTime) bool {
+	aNorm, bNorm := normalize(a, b)
+	return aNorm.Equal(bNorm)
+}
+
+// BeforeOrEqual returns true if the normalized version of a is before or equal to the normalized version of b.
+//
+// Values are normalized to the lowest granularity between the two values: seconds if either
+// is metav1.Time, microseconds if either is metav1.MicroTime, or monotonically-stripped if
+// both are time.Time.
+func BeforeOrEqual(a, b commonTime) bool {
+	aNorm, bNorm := normalize(a, b)
+	return aNorm.Before(bNorm) || aNorm.Equal(bNorm)
+}
+
+// AfterOrEqual returns true if the normalized version of a is after or equal to the normalized version of b.
+//
+// Values are normalized to the lowest granularity between the two values: seconds if either
+// is metav1.Time, microseconds if either is metav1.MicroTime, or monotonically-stripped if
+// both are time.Time.
+func AfterOrEqual(a, b commonTime) bool {
+	aNorm, bNorm := normalize(a, b)
+	return aNorm.After(bNorm) || aNorm.Equal(bNorm)
+}
+
+// normalize returns versions of a and b truncated to the lowest available granularity.
+//
+// 	* If either is metav1.Time, a and b are truncated to time.Second.
+// 	* If either is metav1.MicroTime, a and b are truncated to time.Microsecond.
+// 	* If both a and b are time.Time, a and b have their monotonic clock reading stripped but are otherwise untouched.
+// 	* Otherwise, this function will panic.
+func normalize(a, b commonTime) (time.Time, time.Time) {
+	_, aSecondsGranularity := a.(metav1.Time)
+	_, bSecondsGranularity := b.(metav1.Time)
+	if aSecondsGranularity || bSecondsGranularity {
+		return a.Truncate(time.Second), b.Truncate(time.Second)
+	}
+
+	_, aMicrosecondsGranularity := a.(metav1.MicroTime)
+	_, bMicrosecondsGranularity := b.(metav1.MicroTime)
+	if aMicrosecondsGranularity || bMicrosecondsGranularity {
+		return a.Truncate(time.Microsecond), b.Truncate(time.Microsecond)
+	}
+
+	// this will need new cases if a new time type is ever introduced
```

Comment:
it's not super likely that we'll introduce a new time type anytime soon, but this will be a PITA to find and update if that every happens -- how annoying would it be to do this as a type switch? e.g.:
```
var anySeconds bool
var anyMicroseconds bool
for _, t := range []commonTime{a, b} {
    switch entity.Obj.(type) {
    case metav1.Time:
        anySeconds = true
    case metav1.MicroTime:
        anyMicroseconds = true
    case time.Time:
    default:
        panic("omgwtf")
}

if anySeconds {
    return a.Truncate(time.Second), b.Truncate(time.Second)
} else if anyMicroseconds...
```

Reply:
Oooo thank you, this was a great suggestion (especially given what Nick pointed out re: pointer types). I am embarrassed by my previous logic now 🙈